migration/src/m0000475_improve_version_comparison_fns/semver_cmp.sql

```diff
@@ -0,0 +1,128 @@
+    right_numeric bool;
+
+    cur integer;
+
+begin
+
+    raise notice 'semver_cmp % %', left_p, right_p;
+
+    left_parts = regexp_split_to_array(left_p, E'\\+');
+    left_build = left_parts[2];
+
+    left_parts = regexp_split_to_array(left_parts[1], E'-');
+    left_pre = left_parts[2];
+
+    left_parts = regexp_split_to_array(left_parts[1], E'\\.');
+    left_major = left_parts[1]::decimal;
+    left_minor = left_parts[2]::decimal;
+    left_patch = left_parts[3]::decimal;
+
+    right_parts = regexp_split_to_array(right_p, E'\\+');
+    right_build = right_parts[2];
+
+    right_parts = regexp_split_to_array(right_parts[1], E'-');
+    right_pre = right_parts[2];
+
+    right_parts = regexp_split_to_array(right_parts[1], E'\\.');
+    right_major = right_parts[1]::decimal;
+    right_minor = right_parts[2]::decimal;
+    right_patch = right_parts[3]::decimal;
+
+    if left_major > right_major then
+        return +1;
+    elsif left_major < right_major then
+        return -1;
+    end if;
+
+    if left_minor > right_minor then
+        return +1;
+    elsif left_minor < right_minor then
+        return -1;
+    end if;
+
+    if left_patch > right_patch then
+        return +1;
+    elsif left_patch < right_patch then
+        return -1;
+    end if;
+
+    if left_pre is null and right_pre is not null then
+        return +1;
+    elsif left_pre is not null and right_pre is null then
+        return -1;
+    elsif left_pre is not null and right_pre is not null then
+        left_parts = regexp_split_to_array(left_pre, E'\\.');
+        right_parts = regexp_split_to_array(right_pre, E'\\.');
+        -- do the hard work
+
+        cur := 0;
+        loop
+            cur := cur + 1;
+
+            left_pre := left_parts[cur];
+            right_pre := right_parts[cur];
+
+            if left_pre is null and right_pre is null then
+                return 0;
+            end if;
+
+            if left_pre is null and right_pre is not null then
+                return -1;
+            elsif left_pre is not null and right_pre is null then
+                return +1;
+            end if;
+
+            left_numeric := is_numeric(left_pre);
+            right_numeric := is_numeric(right_pre);
+
+            if left_numeric and right_numeric then
+                if left_pre::bigint < right_pre::bigint then
+                    return -1;
+                elsif left_pre::bigint > right_pre::bigint then
+                    return +1;
+                end if;
+            else
+                if left_pre < right_pre then
+                    return -1;
+                elsif left_pre > right_pre then
+                    return +1;
+                end if;
+            end if;
+
+            if cur > 10 then
+                exit;
+            end if;
+        end loop;
+    else
+        return 0;
+    end if;
+
+    return null;
+exception
+    when others then
+        return null;
+end
+$$
+    language 'plpgsql';
```

Comment:
have u considered adding IMMUTABLE PARALLEL SAFE; here ?

Reply:
Considering that I've no idea what that means... no. Should I?

Reply:
as this function does no mutation, and only works on 2 supplied inputs ... I would suggest it ... usually results in much faster function.